sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```diff
@@ -310,8 +312,19 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
          |$evalSubexpr
          |$writeExpressions
        """.stripMargin
+
+    val runInsideLoop = expressions.exists {
+      case _: LambdaVariable => true
+      case _ => false
+    }
+    val extractValueCode = if (runInsideLoop) {
+      s"$rowWriter.getRow().copy()"
+    } else {
+      s"$rowWriter.getRow()"
+    }
```

Comment:
Nice solution! Correct me if I'm wrong, but I believe this means it should be safe to get rid of the top-level copy from applying `makeCopyIfInstanceOf` in `MapObjects`

Reply:
I'm not an expert on SQL area (limited knowledge) so I'm not sure this is only one path - I'd feel safe if experts in SQL area jump in and review the approach. If we are uncertain and don't feel safe to fix for discovered case, I may need to make more change.

Reply:
Agree that we should fix `CodegenContext.generateExpressionAndCopyIfUnsafe`. To simplify codegen, maybe we can add `copyUnsafeData` to `InternalRow`/`ArrayData`/`MapData`.

Reply:
Just to follow up review comment clearly, there's no `CodegenContext,generateExpressionAndCopyIfUnsafe`. Maybe you referred to `makeCopyIfInstanceOf` inside of `MapObjects.doGenCode`, or suggest to create a new method `CodegenContext,generateExpressionAndCopyIfUnsafe`?